tests/e2e/ctl_v3_move_leader_test.go

```diff
@@ -117,3 +125,82 @@ func testCtlV3MoveLeader(t *testing.T, cfg etcdProcessClusterConfig) {
 		}
 	}
 }
+
+func testCtlV3MoveLeaderWithEnv(t *testing.T, cfg etcdProcessClusterConfig) {
+	defer testutil.AfterTest(t)
+
+	epc := setupEtcdctlTest(t, &cfg, true)
+	defer func() {
+		if errC := epc.Close(); errC != nil {
+			t.Fatalf("error closing etcd processes (%v)", errC)
+		}
+	}()
+
+	var tcfg *tls.Config
+	if cfg.clientTLS == clientTLS {
+		tinfo := transport.TLSInfo{
+			CertFile:      certPath,
+			KeyFile:       privateKeyPath,
+			TrustedCAFile: caPath,
+		}
+		var err error
+		tcfg, err = tinfo.ClientConfig()
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	var leadIdx int
+	var leaderID uint64
+	var transferee uint64
+	for i, ep := range epc.EndpointsV3() {
+		cli, err := clientv3.New(clientv3.Config{
+			Endpoints:   []string{ep},
+			DialTimeout: 3 * time.Second,
+			TLS:         tcfg,
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		resp, err := cli.Status(ctx, ep)
+		if err != nil {
+			t.Fatalf("failed to get status from endpoint %s: %v", ep, err)
+		}
+		cancel()
+		cli.Close()
+
+		if resp.Header.GetMemberId() == resp.Leader {
+			leadIdx = i
+			leaderID = resp.Leader
+		} else {
+			transferee = resp.Header.GetMemberId()
+		}
+	}
+
+	os.Setenv("ETCDCTL_API", "3")
+	defer os.Unsetenv("ETCDCTL_API")
+	cx := ctlCtx{
+		t:           t,
+		cfg:         *newConfigNoTLS(),
+		dialTimeout: 7 * time.Second,
+		epc:         epc,
+		envMap:      map[string]struct{}{},
+	}
+
+	tests := []struct {
```

Comment:
If we need to have different testcases for different tests... let's expose tests as this method parameter.

Reply:
I think we should have exact same test cases. I am executing:
```
func TestCtlV3MoveLeaderSecure(t *testing.T) {
	testCtlV3MoveLeader(t, withCfg(*newConfigTLS()))
	testCtlV3MoveLeader(t, withCfg(*newConfigTLS()), withFlagByEnv())
}

func TestCtlV3MoveLeaderInsecure(t *testing.T) {
	testCtlV3MoveLeader(t, withCfg(*newConfigNoTLS()))
	testCtlV3MoveLeader(t, withCfg(*newConfigNoTLS()), withFlagByEnv())
}
```

Move leader with and without env, which is correct way we need to test both. At the moment I am only fighting with:
```
	{
		{ // request to non-leader
			ret.prefixArgs([]string{ret.epc.EndpointsV3()[(leadIdx+1)%3]}),
			"no leader endpoint given at ",
		},
		{ // request to leader
			ret.prefixArgs([]string{ret.epc.EndpointsV3()[leadIdx]}),
			fmt.Sprintf("Leadership transferred from %s to %s", types.ID(leaderID), types.ID(transferee)),
		},
	}
```

The first test sets the leader to e.g. 2010 and connects to 2005, for example, so it should fail with:
```
no leader endpoint given at ",
```
But the second sets it to point to leader, with the CmdArgs it works, but with Env, it just gets overwritten and the first test is executed with Envs from the second test. I am trying to find out how to prevent that overwrites.

Reply:
Whenever tc.prefixArgs is called it sets an env.

During populating tests struct it overwrites the Env.

Reply:
https://github.com/etcd-io/etcd/blob/master/tests/e2e/ctl_v3_test.go#L283

Reply:
I think the best would be to add env into each test struct, what do you think @ptabor  ?